### PR TITLE
cpu/stm32f1: modernized peripheral GPIO driver

### PR DESCRIPTION
**Pull request by haukepetersen**
One more CPU on the bright side :-)

Code size before (`tests/periph_gpio`, `iot-lab_M3`):
```
   text	   data	    bss	    dec	    hex
  16492	    120	   2888	  19500	   4c2c
```
and after:
```
   text	   data	    bss	    dec	    hex
  15244	    120	   2888	  18252	   474c
```
**Savings ~1,2k**

### CODE REVIEW

**Comment by kaspar030** [comment]
nice!

File: cpu/stm32f1/include/periph_cpu.h [hunk 1]
```diff
@@ -19,13 +19,98 @@
...
+#define HAVE_GPIO_FLANK_T
+typedef enum {
+    GPIO_FALLING = 2,       /**< emit interrupt on falling flank */
+    GPIO_RISING = 1,        /**< emit interrupt on rising flank */
+    GPIO_BOTH = 3           /**< emit interrupt on both flanks */
+} gpio_flank_t;
+/** @} */
+
+/**
+ * @brief   Available ports on the STM32F4 family
```
Comment (PeterKietzmann) [inline cpu/stm32f1/include/periph_cpu.h]: stm32f**1** family

File: cpu/stm32f1/include/periph_cpu.h [hunk 2]
```diff
@@ -19,13 +19,98 @@
...
+    GPIO_AF11,              /**< use alternate function 11 */
+    GPIO_AF12,              /**< use alternate function 12 */
+    GPIO_AF13,              /**< use alternate function 13 */
+    GPIO_AF14               /**< use alternate function 14 */
+} gpio_af_t;
+
+/**
+ * @brief   Configure the alternate function for the given pin
+ *
+ * @note    This is meant for internal use in STM32F4 peripheral drivers only
```
Comment (PeterKietzmann) [inline cpu/stm32f1/include/periph_cpu.h]: stm32f**1** family

File: cpu/stm32f1/periph/gpio.c [hunk 3]
```diff
@@ -30,1015 +30,190 @@
...
+
 typedef struct {
     gpio_cb_t cb;       /**< callback called from GPIO interrupt */
     void *arg;          /**< argument passed to the callback */
-} gpio_state_t;
+} exti_ctx_t;
 
-static gpio_state_t config[GPIO_NUMOF];
+/**
+ * @brief   Allocate memory for on callback and argument per EXTI channel
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: This reads strange somehow. Maybe remove "on" or add a minus, e.g.: "on-callback"

File: cpu/stm32f1/periph/gpio.c [hunk 4]
```diff
@@ -30,1015 +30,190 @@
...
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
 
+
+    /* enable the clock for the selected port */
+    RCC->APB2ENR |= (RCC_APB2ENR_IOPAEN << _port_num(pin));
+    /* clear configuration */
+    port->CR[pin_num >> 3] &= ~(0xf << ((pin_num & 0x7) * 4));
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: Probably I'm just at loss currently. I'm wondering why you shift by `3` and not by `4` here. You want to distinguish between the `CR-low` and the `CR-high` registers (by indexing with `0` or `1`). And AFAIK you want to have `CR[0]` if `pin_num < 16` and  `CR[1]` if `pin_num >=16`. Now just two examples:

```
pin_num = 0d15 = 0b1111
pin_num >> 3 = 0d1 = 0b0001
```

```
pin_num = 0d16 = 0b10000
pin_num >> 3 = 0d2 = 0b0010
```

Where am I mistaken?

Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: IIRC it's the same reason why you do `&0x7` and I think we already had that discussions but I forgot the outcome. Sorry...

File: cpu/stm32f1/periph/gpio.c [hunk 5]
```diff
@@ -30,1015 +30,190 @@
...
-            return;
-    }
-
-    /* unmask the pins interrupt channel */
-    EXTI->IMR |= (1 << exti_line);
-
-    return;
+    (void)pin;
+    (void)af;
+    /* TODO: implement and use this in other peripheral drivers */
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: Is this dependent for other peripheral drivers or didn't you want to implement this function right now :-) ?

File: cpu/stm32f1/periph/gpio.c [hunk 6]
```diff
@@ -30,1015 +30,190 @@
...
-#endif
-        default:
-            return;
-    }
-
-    /* unmask the pins interrupt channel */
-    EXTI->IMR &= ~(1 << exti_line);
-
-    return;
+    EXTI->IMR |= (1 << _pin_num(pin));
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: Nice reduction of code lines!

File: cpu/stm32f1/periph/gpio.c [hunk 7]
```diff
@@ -30,1015 +30,190 @@
...
+        gpio_clear(pin);
     }
     else {
-        gpio_set(dev);
+        gpio_set(pin);
     }
 }
 
-void gpio_write(gpio_t dev, int value)
+void gpio_write(gpio_t pin, int value)
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: I never really understood why we need this function.. Anyway, it's not related to this PR

**Comment by PeterKietzmann** [comment]
The changes look valid to me at a first glance. I don't have all boards here to test, just the iot-lab_M3. Will do now

**Comment by PeterKietzmann** [comment]
The transceiver driver seems so work correctly on the iot-lab_M3 node. Does anyone know the pin-setup of the [iot-lab connector](github.com/iot-lab/iot-lab/wiki/Docs/openm3-schematics.pdf)? Can anyone test this on spark-core and fox?

File: cpu/stm32f1/periph/gpio.c (hunk 4 above)
Comment (haukepetersen) [inline cpu/stm32f1/periph/gpio.c]: you are almost right. I but the CPU only has 16 pins per port, so I want to distinguish between `x < 8` and `x >= 8`... So look at this:
```
pin_num = 0xd7 = 0b0111
pin_num >> 3 = 0d0 = 0b0000
```
```
pin_num = 0d8 = 0b1000
pin_num >> 3 = 0d1 = 0b001
```

File: cpu/stm32f1/periph/gpio.c (hunk 5 above)
Comment (haukepetersen) [inline cpu/stm32f1/periph/gpio.c]: The correct comment must be "use this in other peripheral drivers"...

File: cpu/stm32f1/periph/gpio.c (hunk 7 above)
Comment (haukepetersen) [inline cpu/stm32f1/periph/gpio.c]: this is very useful when setting a pin value programmatical (is this an English word?!) - see #3538 for example...

File: cpu/stm32f1/periph/gpio.c (hunk 4 above)
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/gpio.c]: ahja, remember it :-)

**Comment by PeterKietzmann** [comment]
Ok, how do we proceed? @haukepetersen did you test it on spark-core and/or fox boards? We don't have those...

**Comment by haukepetersen** [comment]
I will try it out and fix the comments tomorrow.

**Comment by haukepetersen** [comment]
addressed comments and added missing implementation of the `init_af` function. Will make a follow up PR to adjust the other peripheral drivers accordingly.

**Comment by haukepetersen** [comment]
I am un-able to test the spark-core, my board seems to be electrically broken...

**Comment by haukepetersen** [comment]
found one more bug in the `_port_pin()` function, a `-2` offset was missing...

**Comment by PeterKietzmann** [comment]
Why is this offset needed? 
I tested this PR again on iot-lab_M3 with *tests/driver_at86rf2xx*  and *tests/periph_gpio*. It seems to be fine.

**Comment by haukepetersen** [comment]
Without the offset, Port A would lead to `2` -> which breaks the clock initialization. This just did not show, as the clocks were also activated hard-coded by other peripheral drivers...

**Comment by haukepetersen** [comment]
two more fixes: 
- removed the GPIO_xx defines from the iot-labs periph_conf
- enable GPIO clock in `gpio_init_af` function

File: cpu/stm32f1/include/periph_cpu.h [hunk 8]
```diff
@@ -19,13 +19,88 @@
...
+
+/**
+ * @brief   Override values for pull register configuration
+ * @{
+ */
+#define HAVE_GPIO_PP_T
+typedef enum {
+    GPIO_NOPULL = 4,        /**< do not use internal pull resistors */
+    GPIO_PULLUP = 9,        /**< enable internal pull-up resistor */
+    GPIO_PULLDOWN = 8       /**< enable internal pull-down resistor */
```
Comment (PeterKietzmann) [inline cpu/stm32f1/include/periph_cpu.h]: Could you shortly explain these values? Maybe referring to the [reference manual](http://www.st.com/st-web-ui/static/active/en/resource/technical/document/reference_manual/CD00171190.pdf) (sec.9.2.1, p. 171 or table 20, p.161)? I think it's much faster than figuring this out by myself

Comment (haukepetersen) [inline cpu/stm32f1/include/periph_cpu.h]: Sure: To understand them, have a look at the gpio_init function and how they are used. I made a bit-map table with the configuration values for the CR[L|H] register and then tried to map the values of DIR and PP to this table. The values you see here is what you get.

Comment (PeterKietzmann) [inline cpu/stm32f1/include/periph_cpu.h]: Ah, I totally overlooked ` if (dir == GPIO_DIR_OUT)` there

**Comment by PeterKietzmann** [comment]
@haukepetersen on iot-lab_M3 I tested tests/periph_gpio, tests/driver_at86rf2xx, examples/default with success. Travis gives us green light. I think no one else will test this on fox and spark-core. So this is an ACK. Please squash your latest commits (or at least remove the "s")

**Comment by haukepetersen** [comment]
squashed. Waiting for Travis once more to make sure I did not break anything while squashing...

**Comment by PeterKietzmann** [comment]
Let's do this. ACK and go

**Pull request merged**